src/ConstructionBase.jl

```diff
@@ -101,21 +115,14 @@ tuple_or_ntuple(::Type, names, vals) = error("Only Int and Symbol propertynames
 
 if VERSION >= v"1.7"
     function getproperties(obj)
-        fnames = propertynames(obj)
-        tuple_or_ntuple(fnames, getproperty.((obj,), fnames))
-    end
-    function getfields(obj::T) where {T}
-        fnames = fieldnames(T)
-        NamedTuple{fnames}(getfield.((obj,), fnames))
-    end
-else
-    @generated function getfields(obj)
-        fnames = fieldnames(obj)
-        fvals = map(fnames) do fname
-             Expr(:call, :getfield, :obj, QuoteNode(fname))
+        if properties_are_fields(obj)
```

Comment:
Indeed; shall we use `is_propertynames_overloaded` on all Julia versions instead to check whether we can safely use fields instead of properties?

Reply:
Using `which` inside a generated function could cause some problems

Reply:
> Using which inside a generated function could cause some problems

Indeed, we hit the infamous: `ERROR: code reflection cannot be used from generated functions` on recent Julia versions.

Not using a generated function greatly decreases the performance, because `which` has to be called at runtime. Unless you know of a way to use such a check in a fast way (tried `hasmethod` which is fast on 1.10+, it does not differentiate between a fallback and a more specific method), I suggest to leave things as they are.

Arguably, overloading `propertynames` with exactly the same set of names as `fieldnames` and have `getproperty` return different values/types is shady enough to justify allowing that edge case to fail, if we don't have a performant way to support it.

Reply:
In particular, I'd suspect the user wouldn't bother implementing `Base.propertynames` if they overload `Base.getproperty` but keep the same set of names, so using that `is_propertynames_overloaded` check would likely fail in this case too.

Reply:
Right, I can see how this affects types that have programmatically-defined `getproperties`/`setproperties` in case of clashes with the underlying data fields.

Do you have other examples in mind? To have a better idea of how "costly" it would be to have to drop support these edge cases. I see that StructArrays.jl has `ConstructionBase` as a direct dependency, so this case may be solved with overloading `ConstructionBase.getproperties`. I can make a PR to do that if we intend to merge this one.

From what I understand, this PR drops support for these edge cases, but provides a significant performance improvement for a large category of types; It's a trade-off but I hope a nice benefit overall.

Reply:
You've convinced me, this looks like a big footgun indeed.

> Maybe worth asking on slack/discourse/... what's the best way to implement getproperties in the general case?

Good idea.

If we don't find a nice way to do it, we could revert `getproperties` as it was; especially now that `setproperties` no longer depends on `getproperties`. We'd still fix the original issue #55 and Accessors/Setfield-related functions would gain the performance improvement without the additional footgun on `getproperties`.

I pushed a new change in https://github.com/JuliaObjects/ConstructionBase.jl/pull/91/commits/24b59df33daaee0209f3ab72c360b6bb314651a8 to use `getfield` instead of `getproperty` for `setfields_object`, but I'm not sure which of these two we should actually use, for the reasons pointed out in the above discussion. Any thoughts?

Reply:
I restored the original `getproperties` implementation with some extra union splitting as mentioned on Zulip to get better performance when structs have a small amount of union members.

We should no longer be adding any footguns in this PR (the same one still exists for `setproperties` but that isn't new to this PR).